opm/common/quad.hpp

```diff
@@ -0,0 +1,325 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+/*!
+ * \file
+ *
+ * \brief This file provides the infrastructure to use quad-precision
+ *        floating point values in the numerical models.
+ */
+#if !defined OPM_COMMON_QUAD_HPP && HAVE_QUAD
+#define OPM_COMMON_QUAD_HPP
+
+#ifdef DUNE_CLASSNAME_HH
+#error "Due to some trickery required for the linker, this file must be included _before_ Dune's classname.hh!"
+#endif
+
+#include <dune/common/classname.hh>
```

Comment:
> What is wrong with the ClassName.hpp solution?

it would mean that we need to keep maintaining this file, it will be more intransparent (because you have to include OPM's ClassName.hpp to get `Dune::className()`) and it would be more error-prone (because including the wrong file will silently work for everything *except* if you try to use `Dune::className()` on `__float128` objects)

> I'd think it would mean less disruption, too?

only for code that currently uses OPM's version of ClassName.hpp. since that is limited to opm-material and eWoms I think the overall burden of keeping and maintaining this file is higher.